### PR TITLE
model: support dots.vlm1 model

### PR DESCRIPTION
**Pull request by yonghenglh6**
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/multimodal/processors/dots_vlm.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import math
+import asyncio
+from PIL import Image
+from sglang.srt.models.dots_vlm import DotsVLMForCausalLM
+from sglang.srt.multimodal.processors.base_processor import (
+    BaseMultimodalProcessor,
+    MultimodalSpecialTokens,
+)
+
+class DotsVLMImageProcessor(BaseMultimodalProcessor):
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/dots_vlm.py:16]: Can we reuse Qwen2_5VLImageProcessor for less code change?

File: python/sglang/srt/models/dots_vlm_vit.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.pre_norm:
+            x = self.mlp(self.ln_q(x).view(-1, self.hidden_size))
+        else:
+            x = self.mlp(x.view(-1, self.hidden_size))
+        return x
+
+
+class VisionAttention(nn.Module):
```
Comment (mickqian) [inline python/sglang/srt/models/dots_vlm_vit.py]: use `VisionAttention` in sglang

File: python/sglang/srt/models/dots_vlm_vit.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
+import math
```
Comment (mickqian) [inline python/sglang/srt/models/dots_vlm_vit.py]: Please:
1. use VisionAttention in sglang
2. add quant_config to init
3. add copyright at the top of the model file

**Review by mickqian — CHANGES_REQUESTED** [review]
Since the vit and processor implementation is similar to Qwen2-VL series, we should reuse the previous code as much as possible.

You can ref https://github.com/mickqian/sglang/tree/xhs-vlm

File: python/sglang/srt/conversation.py [hunk 4]
```diff
@@ -967,6 +978,33 @@ def generate_chat_conv(
...
+        system_template="<|system|>{system_message}<|endofsystem|>\n",
+        roles=("<|user|>", "<|assistant|>"),
+        sep="<|endofuser|>",
+        sep2="<|endofassistant|>",
+        stop_str=["<|endofassistant|>"],
+        sep_style=SeparatorStyle.DOTS_VLM,
+        image_token="<|img|><|imgpad|><|endofimg|>",
+    )
+)
+
```
Comment (JustinTong0323) [inline python/sglang/srt/conversation.py:1007]: We don't need to manually add templates now, ideally we would auto use the chat-template provided in hf model files

File: python/sglang/srt/models/dots_vlm_vit.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+            spatial_merge_size=config.spatial_merge_size,
+            init_merger_std=self.config.init_merger_std,
+        )
+
+        self.gradient_checkpointing = False
+        self._gradient_checkpointing_func = torch.utils.checkpoint.checkpoint
+
+    def _init_weights(self, module):
+        std = self.config.initializer_range
+        if isinstance(module, (nn.Linear, nn.Conv3d)):
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/models/dots_vlm_vit.py]: fixed

File: python/sglang/srt/models/dots_vlm_vit.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+        cu_seqlens = F.pad(cu_seqlens, (1, 0), value=0)
+
+        for blk in self.blocks:
+            if self.gradient_checkpointing and self.training:
+                hidden_states = self._gradient_checkpointing_func(
+                    blk.__call__,
+                    hidden_states,
+                    cu_seqlens,
+                    rotary_pos_emb,
+                    use_reentrant=(self.config.ckpt_use_reentrant or self.config.ve_ckpt_use_reentrant),
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/models/dots_vlm_vit.py]: removed

File: python/sglang/srt/configs/dots_vlm.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            for i in range(len(text)):
+                while self.image_token in text[i]:
+                    text[i] = text[i].replace(
+                        self.image_token,
+                        "<|placeholder|>"
+                        * (image_grid_thw[index].prod() // merge_length),
+                        1,
+                    )
+                    index += 1
+                text[i] = text[i].replace("<|placeholder|>", self.image_token)
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/configs/dots_vlm.py]: removed and use transformers.qwen_2_5_processor

File: python/sglang/srt/configs/dots_vlm.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    DeepseekV2Config,
+)
+from typing import Any, Optional, List, Union
+from transformers import (
+    AutoProcessor,
+    LlamaTokenizerFast,
+    PretrainedConfig,
+    ProcessorMixin,
+)
+# from transformers.models.auto.configuration_auto import CONFIG_MAPPING
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/configs/dots_vlm.py]: removed

File: python/sglang/srt/models/dots_vlm_vit.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+    ) -> None:
+        super().__init__()
+        self.hidden_size = context_dim * (spatial_merge_size**2)
+        self.pre_norm = pre_norm
+        if self.pre_norm == "layernorm":
+            self.ln_q = LayerNorm(context_dim, eps=1e-6)
+        elif self.pre_norm == "rmsnorm":
+            self.ln_q = RMSNorm(context_dim, eps=1e-6)
+        else:
+            print("no norm in patch merger")
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/models/dots_vlm_vit.py]: use logger now

File: python/sglang/srt/multimodal/processors/dots_vlm.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        ) -> tuple[int, int]:
+            """Rescales the image so that the following conditions are met:
+
+            1. Both dimensions (height and width) are divisible by 'factor'.
+
+            2. The total number of pixels is within the range ['min_pixels', 'max_pixels'].
+
+            3. The aspect ratio of the image is maintained as closely as possible.
+            """
+            if max(height, width) / min(height, width) > self.MAX_RATIO:
```
[the remark this answers is not in the document]
Comment (redmoe-moutain) [reply python/sglang/srt/multimodal/processors/dots_vlm.py]: use sglang.srt.multimodal.processors.qwen_vl smart_resize now

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 1 above)
Comment (redmoe-moutain) [reply python/sglang/srt/multimodal/processors/dots_vlm.py:16]: thanks for the example  https://github.com/mickqian/sglang/tree/xhs-vlm. I do not use Qwen2_5VLImageProcessor directly, but derive from it and overrides attributes, tokenizer_class and some special tokens; For DotsVLMImageProcessor, I only import smart_resize from qwen_vl, because DotsVLMImageProcessor do not need mrope_positions, video_processor. And we will enable DotsVLMImageProcessor to support dynamic min/max pixels per image soon, it is more convenient to use a standlone DotsVLMImageProcessor.

File: python/sglang/srt/models/dots_vlm_vit.py (hunk 2 above)
Comment (redmoe-moutain) [reply python/sglang/srt/models/dots_vlm_vit.py]: done

File: python/sglang/srt/models/dots_vlm_vit.py (hunk 3 above)
Comment (jianfei-wangg) [reply python/sglang/srt/models/dots_vlm_vit.py]: > Please:
> 
> 1. use VisionAttention in sglang
> 2. add quant_config to init
> 3. add copyright at the top of the model file

I have finished them, please review our code again, thanks.

File: python/sglang/srt/models/dots_vlm.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
+# Copyright 2025 The RedNote HiLab team.
```
Comment (mickqian) [inline python/sglang/srt/models/dots_vlm.py:1]: should we reuse Vit of qwen2_5vl as well?

File: python/sglang/srt/multimodal/processors/dots_vlm.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        ):
+            image_data = sum(image_data, [])
+
+        base_output = self.load_mm_data(
+            prompt=input_text,
+            image_data=image_data,
+            multimodal_tokens=self.mm_tokens,
+        )
+
+        def resize_image(image, size_factor: int = self.IMAGE_FACTOR) -> Image.Image:
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/dots_vlm.py]: reuse the method with the same name in qwen_vl.py

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 1 above)
Comment (mickqian) [reply python/sglang/srt/multimodal/processors/dots_vlm.py:16]: That way, we should probably inherit from `Qwen2_5VLImageProcessor`?

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 12 above)
Comment (jianfei-wangg) [reply python/sglang/srt/multimodal/processors/dots_vlm.py]: `def resize_image(image, size_factor: int = IMAGE_FACTOR) -> Image.Image:
    width, height = image.size
    min_pixels = MIN_PIXELS
    max_pixels = MAX_PIXELS`
qwen_vl use hardcoded MIN_PIXELS and MAX_PIXELS，which is different from ours. Additionally, we will support dynamic resolution per image soon, So I do not use resize_image in qwen_vl.py directly. Can I change code in qwen_vl.py?

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 1 above)
Comment (jianfei-wangg) [reply python/sglang/srt/multimodal/processors/dots_vlm.py:16]: > That way, we should probably inherit from `Qwen2_5VLImageProcessor`?

I'm not quite sure. DotsVLMImageProcessor do not need mrope_positions, video_processor.  Should I make DotsVLMImageProcessor inheritting from Qwen2_5VLImageProcessor and overrides resize_image and process_mm_data_async method?

File: python/sglang/srt/models/dots_vlm.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+    @classmethod
+    def get_model_config_for_expert_location(cls, config):
+        return DeepseekV2ForCausalLM.get_model_config_for_expert_location(config)
+
+    def pad_input_ids(self, input_ids: List[int], mm_inputs: MultimodalInputs):
+        """Pad input_ids with multimodal tokens"""
+        # Get image token ID for padding pattern
+        pattern = MultiModalityDataPaddingPatternMultimodalTokens()
+        mm_inputs.im_token_id = self.image_token_id
```
Comment (JustinTong0323) [inline python/sglang/srt/models/dots_vlm.py]: It's weird, mm_input should contains the im_token_id if we return in processor

File: python/sglang/srt/models/dots_vlm.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+    def get_image_feature(self, items: List[MultimodalDataItem]) -> torch.Tensor:
+        """Extract image features from multimodal data items"""
+        if any(item.precomputed_embeddings is not None for item in items):
+            # If features are precomputed, just concatenate them
+            if not all(item.precomputed_embeddings is not None for item in items):
+                raise NotImplementedError(
+                    "MM inputs where only some items are precomputed."
+                )
+            return torch.concat(
+                [torch.tensor(item.precomputed_embeddings) for item in items]
```
Comment (JustinTong0323) [inline python/sglang/srt/models/dots_vlm.py]: precomputed_embedding is processed in general_mm_embed_routine, no need here.

File: python/sglang/srt/conversation.py (hunk 4 above)
Comment (JustinTong0323) [inline python/sglang/srt/conversation.py]: Remove it. We don't need to manually add templates now, ideally we would auto use the chat-template provided in hf model files (or in tokenizer manager)

File: python/pyproject.toml [hunk 15]
```diff
@@ -13,7 +13,7 @@ classifiers = [
     "Programming Language :: Python :: 3",
     "License :: OSI Approved :: Apache Software License",
 ]
-dependencies = ["aiohttp", "requests", "tqdm", "numpy", "IPython", "setproctitle"]
+dependencies = ["aiohttp", "requests", "tqdm", "numpy", "IPython", "setproctitle", "sentencepiece"]
```
Comment (JustinTong0323) [inline python/pyproject.toml]: why we need this?

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 12 above)
Comment (mickqian) [reply python/sglang/srt/multimodal/processors/dots_vlm.py]: Yes, we can add parameters to it

File: python/pyproject.toml (hunk 15 above)
Comment (jianfei-wangg) [reply python/pyproject.toml]: reverted

File: python/sglang/srt/models/dots_vlm.py (hunk 13 above)
Comment (jianfei-wangg) [reply python/sglang/srt/models/dots_vlm.py]: yes, removed

File: python/sglang/srt/models/dots_vlm.py (hunk 14 above)
Comment (jianfei-wangg) [reply python/sglang/srt/models/dots_vlm.py]: removed

File: python/sglang/srt/conversation.py (hunk 4 above)
Comment (jianfei-wangg) [reply python/sglang/srt/conversation.py]: removed

File: python/sglang/srt/multimodal/processors/dots_vlm.py (hunk 12 above)
Comment (jianfei-wangg) [reply python/sglang/srt/multimodal/processors/dots_vlm.py]: I have added parameters to qwen-resize_image, and then import it

File: python/sglang/srt/models/dots_vlm.py (hunk 11 above)
Comment (jianfei-wangg) [reply python/sglang/srt/models/dots_vlm.py:1]: The two are not exactly the same, and the algorithm engineer has not aligned them yet

**Review by mickqian — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
